spec/controllers/orders_controller_spec.rb

```diff
@@ -0,0 +1,156 @@
+require "rails_helper"
+include SessionsHelper
+include CartsHelper
+
+RSpec.describe OrdersController, type: :controller do
+  describe "GET #new" do
+    let(:user) {FactoryBot.create :user}
+    let(:product) {FactoryBot.create :product}
+    let(:delivery_addresses) {FactoryBot.create :delivery_address}
+
+    context "success when valid attributes" do
+      before do
+        log_in user
+        session[:cart] = {Product.first.id.to_s => 1}
+        get :new
+      end
+
+      it "assign @delivery_addresses" do
+        expect(assigns(:delivery_addresses)).not_to be_nil
+      end
+
+      it "assign @products_load_in_cart" do
+        expect(assigns(:products_load_in_cart)).not_to be_nil
+      end
+
+      it {should render_template("new")}
+    end
+
+    context "fail when user login not yet" do
+      before do
+        get :new
+      end
+
+      it "is store forwarding_url" do
+        expect(session[:forwarding_url]).not_to be_nil
+      end
+
+      it "show flash error" do
+        expect(flash[:error]).to eq I18n.t("not_log_in")
+      end
+
+      it {should redirect_to(login_path)}
+    end
+
+    context "when cart empty" do
+      before do
+        log_in user
+        session[:cart] = nil
+        get :new
+      end
+
+      it "show flash warning" do
+        expect(flash[:warning]).to eq I18n.t("cart_empty")
+      end
+
+      it {should redirect_to(carts_path)}
+    end
+  end
+
+  describe "POST #create" do
+    let(:user) {FactoryBot.create :user}
+    let!(:product) {FactoryBot.create :product}
+    let!(:order_params) {FactoryBot.build(:order).as_json}
```

Comment:
code này lặp, -> đưa ra khỏi describe để dùng chung nhé

ngoài ra, sao lại dùng `let(:user` chị dùng `let!` thì có j khác biệt

Reply:
dạ em chuyển sang dùng let! rồi chị, mới đầu em test thì nó giống nhau cả nhưng mà sau khi đọc lại thì em thấy dùng let! thì đúng hơn phải không chị nhỉ, vì nó được run trước ở before rồi mới vài it :D